### DISCUSSION TITLE
Reverse mode in flow models (FastFlow) is never called

### DISCUSSION THREAD

**Discussion by kshiring** [open-edge-platform/anomalib#643] [Q&A] [answered]
Hi everyone, I'm currently doing a deep dive on the FastFlow model in anomalib. In particular, I'm looking at the behavior of the forward() function in the AllInOneBlock module:
```def forward(self, x, c=[], rev=False, jac=True):```
The variable ```rev``` never seems to be set to ```True```. I find this strange, given that the FastFlow paper (and all flow-based models) use reversible functions to compute FF^-1. I would expect that ```rev=True``` would be needed for model training in order to compute the module chain for the embedding -> image prior. However, I see that the training code provided by anomalib never seems to use it.
I have used the ```train.py``` and ```test.py``` tools, and neither seems to trigger the reverse-mode call to ```forward()```. Can someone help me understand why?

**Comment by ashwinvaidya17** [comment, marked as answer]
@kshiring sorry for the late reply.

FastFlow and similar models use the concept of invertible transformations for modelling the source probability distribution. The idea behind normalizing flow is to take a source distribution and map it to a normal distribution using a series of invertible operations. Once such a series of invertible operations is learnt, you can then generate a sample in the source distribution space using a sample from the normal distribution and reversing the flow. This is how normalizing flows are used in generative modeling applications. So going from the target distribution to normal distribution is called the normalizing direction and the opposite is called the generative direction.

The `rev` (or the reverse) parameter of `AllInOneBlock` allows us to either get the result of the normalizing direction (`rev=False`) or the generative direction (`rev=True`). When using normalizing flow to detect the anomalies, FastFlow is only interested in learning a function that can map the source distribution to a normal distribution. So it is only interested in the normalizing direction. During training you maximize the log likelihood (or minimize the negative log likelihood). Hence, the gradients do not depend on the values returned when `rev=True`. This is why we set `rev=False` is never True in the `AllInOneBlock` module. Once we have a good function to map the input image to the normal distribution, any image that is not part of the source distribution will have low likelihood to be from the normal distribution. FastFlow uses 2D normalizing flows to preserve the spatial locations of the embeddings so that we can get likelihoods in two dimensions. We can then use these likelihoods to get the anomaly maps.

Here are two sentences from the paper that might make it clear:

> In training stage, our FastFlow is trained with normal images to transform the original distribution to a standard normal distribution in a 2D manner. In inference, we use the probability value of each location on the two-dimensional feature as the anomaly score.

> In inference, the features of anomalous images should be out of distribution and hence have lower likelihoods than normal images and the likelihood can be used as the anomaly score. Specifically, we sum the two-dimensional probabilities of each channel to get the final map and up sample it to the input image resolution using bilinear interpolation.

Let me know if something is not clear.
